### ISSUE TITLE
Update KCT config for latest KCT

### ISSUE THREAD

**Issue by NathanKell** [KSP-RO/RP-0#122]
@magico13 has released a new, awesome way to configure KCT. We should use it.

**Comment by magico13** [comment]
It's not quite released yet, but should be soon. A bit more testing is needed before I throw it all out into the wild. I'm certainly willing to help you out with any questions and there's still time to make a few more changes if you need or want them (frankly, you guys are the primary people who will take advantage of the features, as most normal users won't want to tinker with all the formula. Though I did have someone asking for some features for a challenge he was working on setting up).

I'm not sure if you noticed, but I have a small writeup about the formula configs on the [KCT github wiki](https://github.com/magico13/KCT/wiki/KCT_Formula.cfg-and-Math-Parser-Documentation). It covers what functions are available and what variables are exposed for each formula. For several formula, a negative result disables it (but 0 usually doesn't). The sign() function can then be used to set the function negative after a certain point (sign(x) returns +1 for x>=0, and -1 for x<0). The default build rate formula uses this to limit the maximum number of rates you can have based on the VAB/SPH level. The "simple" config I posted in the [KCT dev thread](http://forum.kerbalspaceprogram.com/threads/69310-Kerbal-Construction-Time-Dev-Thread?p=1785879&viewfull=1#post1785879) also uses it for a similar function.

You should be able to nearly completely tailor KCT to work how you want. The "simple" config provides an example setup that doesn't even use the KCT upgrade system and only relies on the Stock building upgrades.

Like I mentioned before, if there's any functions you want added or additional formula you want exposed, let me know and I can add them in :) I've been considering adding a one-line if statement (like "if(x > 2?-1:1)", which isn't quite how most languages do them, but would be easier for me write up), but the tricks with sign() can usually work in much the same way.

**Comment by magico13** [comment]
Just to let you know, I'm trying to get a new feature out as soon as I can (but progress has been slower than I would like) where mods can ship full sets of configs without overwriting anything (called Presets). The Preset spec is still a work in progress, but you can take a look at the [Preset class](https://github.com/magico13/KCT/blob/presets/Kerbal_Construction_Time/KCT_PresetManager.cs#L182) to get an idea of what's available. Basically, all the current game specific settings, time related settings, and formulas are being merged into a single file and every save will have their own settings. Presets will be found automatically if a KCT_Presets folder is located in the main mod folder (ie, GameData/YourModFolder/KCT_Presets) meaning you won't need to ship the KCT settings in a separate folder and you won't have to overwrite anything.

When the player starts a new game they choose a Preset and those settings are then applied. If the original Preset gets altered later (for balance purposes probably), the changes will automatically be applied. If the player changes any settings though, it switches to a "Custom" preset which won't inherit changes. You can also change Presets after the save has been created, and can save your settings as a new Preset (well, when I finish that feature).

I've also opened up quite a few more formulas since RP-0 last updated the config. Some are new to the Presets release (simulation times, costs for resetting upgrade point distribution) and aren't fully implemented yet. Some other new settings/features include setting the number of starting upgrade points, disabling getting upgrade points from unlocking the tech tree (for use later with the API so other mods can give points for other achievements, or to require purchasing upgrades), and disabling simulations. I also added the variable [S] to the BuildRateFormula which represents the total number of tech nodes unlocked.

A lot of this is still in progress at the moment, but I hope to get a lot of it finished this weekend. Some settings haven't actually been implemented and the GUI only works so far as to let you select Presets (but not really modify the settings), but loading Presets from other folders is working great and I hopefully won't be changing the specs too much more (I may or may not add another setting or formula).

Now's also a great time to let me know if there's anything else you want exposed in the Presets since nothing is finalized ;)

TL;DR: I'm changing up the KCT configs to use Presets that can be included in your own mod folder and doesn't overwrite anything. There are a lot of options available. It's a mandatory change, but might take me some time to finish. Let me know if you want anything else exposed :)

**Comment by magico13** [comment]
As a starting point for this, I've converted the old RP-0 config to the new system for you. Take the code below and plop it into a file called RP0.cfg (doesn't HAVE to be called that, you could call it KCT.cfg if you wanted, but it must end in ".cfg") into GameData/RP-0/KCT_Presets/RP0.cfg. KCT will find the KCT_Presets folder in RP-0 and will load anything in there that ends with .cfg as a Preset. You won't have to overwrite any KCT settings now, and you won't have to ship a KerbalConstructionTime folder anymore.

The following config is just a rehash of the old one. You'll likely want to take advantage of the new formulas yourself. When I've got some better documentation up I'll link it here, but for now you can look [here](https://github.com/magico13/KCT/wiki/KCT_Formula.cfg-and-Math-Parser-Documentation) for some older documentation. There's a slight chance I'll add or change one or more settings, but I think most things are final at this point.

There's a way to edit all of these in game, but with the allowDeletion flag set to false you won't be able to overwrite it. Any Presets saved in-game are saved in the KerbalConstructionTime/KCT_Presets folder.

````
KCT_Preset
{
	name = KCT Settings for RP-0
	shortName = RP0
	description = The recommended config for using KCT with RP-0
	author = RP-0 Devs
	allowDeletion = False
	career = True
	science = True
	sandbox = True
	KCT_Preset_General
	{
		Enabled = True
		BuildTimes = True
		ReconditioningTimes = True
		TechUnlockTimes = True
		KSCUpgradeTimes = True
		Simulations = True
		SimulationCosts = True
		RequireVisitsForSimulations = True
		TechUpgrades = True
		StartingPoints = 15,15,45
	}
	KCT_Preset_Time
	{
		OverallMultiplier = 25
		BuildEffect = 1
		InventoryEffect = 100
		ReconditioningEffect = 576
		MaxReconditioning = 13500000
		RolloutReconSplit = 0.5
	}
	KCT_Preset_Formula
	{
		NodeFormula = 2^([N]+1) / 86400 / 200
		UpgradeFundsFormula = min(([N]+1)^0.5 * 10000, 1024000)
		UpgradeScienceFormula = min(2^([N]+2) * 1.0, 512)
		ResearchFormula = [N]*0.5/86400
		EffectivePartFormula = min([C]/([I] + ([B]*([U]+1))), [C])
		ProceduralPartFormula = (([C]-[A]) + ([A]*10/max([I],1))) / max([B]*([U]+1),1)
		BPFormula = ([E]^(1/2))*2000*[O]
		KSCUpgradeFormula = ([C]^(1/2))*1000*[O]
		ReconditioningFormula = min([M]*[O]*[E], [X])
		BuildRateFormula = (([I]+1)*0.05*[N] + max(0.1-[I], 0))*sign(2*[L]-[I]+1)
		SimCostFormula = max([C]/50000 * ([PM]/[KM]) * ([S]/10 + 1) * ([A]/10 + 1) * ([L]^0.5) * 100, 1000)
		KerbinSimCostFormula = max([C]/50000 * ([L]^0.5) * 10, 100)
		UpgradeResetFormula = 2*([N]+1)
		InventorySaleFormula = ([V]+[P] / 10000)^(0.5)
	}
}
````

**Comment by NathanKell** [comment]
@magico13 awesome, thanks! I'll take a look at this tomorrow when awake. :)

**Comment by magico13** [comment]
Just a heads up that the RP-0 config for KCT 1.2 isn't formatted correctly, resulting in all of the Formulas being set to their default values. Below should be the correct values. I've removed the comments just to be sure they don't mess anything up (they shouldn't, but I haven't tested) and changed the reconditioning formula to use *abs([RE]-[S]) which makes the reconditioning split slider actually do something.

The main issue was just that there was an extra KCT_FormulaSettings node inside the KCT_Preset_Formula node.

````
KCT_Preset
{
    name = KCT Settings for RP-0
    shortName = RP0
    description = The recommended config for using KCT with RP-0
    author = RP-0 Devs
    allowDeletion = False
    career = True
    science = True
    sandbox = True
    KCT_Preset_General
    {
        Enabled = True
        BuildTimes = True
        ReconditioningTimes = True
        TechUnlockTimes = True
        KSCUpgradeTimes = True
        Simulations = True
        SimulationCosts = True
        RequireVisitsForSimulations = True
        TechUpgrades = True
        StartingPoints = 15,15,45
    }
    KCT_Preset_Time
    {
        OverallMultiplier = 25
        BuildEffect = 1
        InventoryEffect = 100
        ReconditioningEffect = 576
        MaxReconditioning = 13500000
        RolloutReconSplit = 0.5
    }
    KCT_Preset_Formula
    {
            UpgradeFundsFormula = 10000
            NodeFormula = (5+([N]*1.5)) / 86400 / 365 // 5 science/year, + 1.5 sci per upgrade
            EffectivePartFormula = min([C]/([I] + ([B]*(max([U]*0.125,1)+1))), [C])
            ProceduralPartFormula = (([C]-[A]) + ([A]*10/max([I],1))) / max([B]*(max([U]*0.125,1)+1),1)
            BPFormula = (20000 + (([E]^(3/4))*300))*[O]
            KSCUpgradeFormula = ([C]^(3/4))*200*[O]
            ReconditioningFormula = min(7200+((([M]*[O]*[E])^(1/2))*60), [X])*abs([RE]-[S])
            BuildRateFormula = (1+([L]*0.25))*(([I]+1)*0.05*[N] + max(0.1-[I], 0))*sign(2*[L]-[I]+1)
            SimCostFormula = max([C]/50000 * ([PM]/[KM]) * ([S]/10 + 1) * ([A]/10 + 1) * ([L]^0.5) * 100, 1000)
            KerbinSimCostFormula = max([C]/50000 * ([L]^0.5) * 10, 100)
            InventorySaleFormula = 0
            ResearchFormula = -1
            UpgradeResetFormula = -1
            UpgradeScienceFormula = -1
    }
}

````
